src/core/Global.ttslua

```diff
@@ -283,6 +284,127 @@ function getTokenIndex(token)
   end
 end
 
+function activeRedrawEffect(originParams)
+  redrawData = originParams
+  makeButtonsToRedraw()
+end
+
+function makeButtonsToRedraw()
+  if isTokenXMLActive == true then
+    broadcastToAll("Clear already active buttons first, then try again", "Red")
+    return
+  end
+
+  if #chaosTokens == 0 then
+    broadcastToAll("No tokens found in play area", "Red")
+    return
+  end
+
+  local chaosbag = findChaosBag()
+  local matchingTokensInPlay = {}
+
+  -- determine if only some tokens are able to be returned to the bag
+  for _, token in ipairs(chaosTokens) do
+    local tokenName = getReadableTokenName(token.getName())
+    -- nil handling
+    redrawData.VALID_TOKENS = redrawData.VALID_TOKENS or {}
+    redrawData.INVALID_TOKENS = redrawData.INVALID_TOKENS or {}
+  
+    -- allow valid tokens or not invalid tokens, also allow any token if both lists empty
+    if (redrawData.VALID_TOKENS[tokenName] and #redrawData.INVALID_TOKENS == 0) ~= nil or (isTableEmpty(redrawData.VALID_TOKENS) and not redrawData.INVALID_TOKENS[tokenName]) or
```

Comment:
Should use the `isTableEmpty()` function for the first condition too (and I think the nil comparision should just be for the first part):

```lua
if (redrawData.VALID_TOKENS[tokenName] ~= nil and isTableEmpty(redrawData.INVALID_TOKENS)) ...
```